Doc 12037 vector search beta2 dotnet

ElliotFrancisHunter opened the pull request:
This PR contains the doc copy for the upcoming Vector Search .NET beta 2 release.

It also has some small fly by fixes:

- Removal of British English on existing Vector Search pages
- Reorder of release notes and upgrade links on what's new to reflect order on left hand nav
- A handful of fixed typos

ElliotFrancisHunter committed:
DOC:12037- Update iOS pages to use includes

Changes:
. Changed inline snippets to included snippets
. Changed snippets to updated/fixed snippets from Pasin
. Added tags to code files
. Corrected typo on code filename

TODO:
. Update working with copy to reflect new snippets

ElliotFrancisHunter committed:
DOC-12037: Beta 2 V1

Changes:
. Fixed bug in android docs that redirected to Swift page instead
. Added .Net and Java Desktop pages
. Working with Vector Search - Java page complete
. Added new pages to nav and changed .Net and Java nav structure to the new format
. Updated what's new to reflect changes

TODO:
. Add .Net samples
. Probably add Beta 2 section to what's new
. Update snippets copy to reflect any code changes (iOS)

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-12037-vector-search-beta2

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-12037-vector-search-beta2-dotnet

ElliotFrancisHunter committed:
DOC-12037: Copy fixes

Remove British English and numbers instead of words

ElliotFrancisHunter committed:
DOC-12037: Vector Search ready for review

Change list:

-- Added Vector Search build and install instructions plus AVX2 admonition
-- Added Vector Search specific module attributes for .NET
--  Completed working with vector search page (.NET) to include correct snippets and reflect snippet content and the new color model
-- Updated API Reference link in the nav

ElliotFrancisHunter committed:
DOC-12037: fix snippet tags

Fixed typos in tags and moved them to correctly display context of snippet

ElliotFrancisHunter committed:
DOC-12037: what's new update

Added .Net beta 2 release information. Changed order of release notes and upgrade links to reflect order in the left hand nav

pasin commented on modules/csharp/pages/gs-build.adoc:
```diff
@@ -30,6 +30,31 @@ The document was successfully persisted to the database.
+[source, csharp]
+----
+// For Android applications
+Couchbase.Lite.Support.Droid.Activate(Context);
```
This line is not required for loading the vector search extension but it's required for using .NET on the android platform in general. With the assumption that this is already explained somewhere else, this line can be removed here.

ElliotFrancisHunter replied:
Done!

pasin commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -58,10 +68,38 @@ This is the recommended method of dependency management because it supports the
+* `Couchbase.Lite.VectorSearch.Support.ios`
+* `Couchbase.Lite.VectorSearch.Support.android`
+* `Couchbase.Lite.VectorSearch.Support.NetDesktop`
+* `Couchbase.Lite.VectorSearch.Support.WinUI`
```
I don't think this list is correct. @borrrden Can you confirm?

pasin replied:
As @borrrden reminded me on slack about this, I think the package will just be "Couchbase.Lite.VectorSearch" and there are no Support packages needed to be installed.

ElliotFrancisHunter replied:
Okay I've removed the support packages in this section for Couchbase.Lite.VectorSearch. I'm assuming you still need the Enterprise packages for CBL so have left them as is.

pasin reviewed (commented):
Need @borrrden to confirm about the supported platforms in the vector search package. The section describing the key concepts is very useful and easy to understand.

borrrden commented on modules/csharp/examples/code_snippets/VectorSearch.cs:
```diff
@@ -67,14 +66,14 @@ private void CreateCustomVectorIndexConfig()
-            var database = new Database("my-database");
-
             // tag::vs-create-index[]
+            var database = new Database("my-database");
```
In general I want to leave lines like this out of the snippets. It should be obvious that you need to create a database, and this just adds needless lines to the snippet.

borrrden commented on modules/csharp/examples/code_snippets/VectorSearch.cs:
```diff
@@ -171,11 +170,12 @@ private void QueryUsingVectorMatch()
 
         private void QueryUsingVectorDistance()
         {
+            // tag::vs-use-vector-distance[]
```
Ditto here, move the `var database` line out of the tag.  Also there is no real need to make the `inputColor` variable here, it could simply be used as a literal later.

ElliotFrancisHunter replied:
Done!

borrrden commented on modules/csharp/pages/gs-build.adoc:
```diff
@@ -30,6 +30,31 @@ The document was successfully persisted to the database.
+----
+// For Android applications
+Couchbase.Lite.Support.Droid.Activate(Context);
+Extension.Load(new VectorSearchExtension(Context));
```
Warning:  These will not get checked by the compiler in the way that snippets will.

ElliotFrancisHunter replied:
Done!

borrrden commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -58,10 +68,38 @@ This is the recommended method of dependency management because it supports the
+This is the recommended method of dependency management because it supports the strict version requirement between the core _Couchbase Lite for .Net_ package and its dependent Support library, which comprises:
+
+.`Couchbase.Lite.Enterprise`
+* `Couchbase.Lite.Enterprise.Support.UWP`
```
Remove this, this platform is removed as of 3.2

ElliotFrancisHunter replied:
Done.

borrrden commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -58,10 +68,38 @@ This is the recommended method of dependency management because it supports the
+==== Package Reference
+This is the recommended method of dependency management because it supports the strict version requirement between the core _Couchbase Lite for .Net_ package and its dependent Support library, which comprises:
+
+.`Couchbase.Lite.Enterprise`
```
Does this really need to be duplicated here?  Aren't these install instructions already somewhere else for the main package?  This is making this page overly verbose imo.

ElliotFrancisHunter replied:
I see your point. I've condensed to a couple of lines but will keep the section to make sure Vector Search instructions are distinguishable.

ElliotFrancisHunter committed:
DOC-12037: .NET release notes

ElliotFrancisHunter committed:
DOC-12037: Addressing feedback

Changelist:
INSTALL
- Removed deprecated .UWP package from dependencies
- Removed incorrect packages from Vector Search section
- Condensed and simplified Vector Search dependencies section
- Changed example to a non-deprecated package and added a note about the Activate() function
---
BUILD
- Removed unnecessary snippet

ElliotFrancisHunter committed:
Merge branch 'release/3.2' into DOC-12037-vector-search-beta2-dotnet

ElliotFrancisHunter committed:
DOC-12037: Final feedback

ElliotFrancisHunter committed:
DOC-12037: Fix missing curly brace

lamagnifica commented on modules/ROOT/pages/cbl-whatsnew.adoc:
```diff
@@ -17,6 +17,27 @@ NOTE: Couchbase Lite 3.0 introduces some breaking changes. +
+--
+Vector Search is available only for 64-bit architectures and
+Intel processors that support the Advanced Vector Extensions 2 (AVX2) instruction set.
+To verify whether your device supports the AVX2 instructions set, https://www.intel.com/content/www/us/en/support/articles/000090473/processors/intel-core-processors.html[follow these instructions.] 
```
instruction singular here?

lamagnifica commented on modules/csharp/pages/gs-build.adoc:
```diff
@@ -30,6 +30,30 @@ The document was successfully persisted to the database.
 
+=== Building With Vector Search
+
+To build your Couchbase Lite application with Vector Search, you must add the following to your existing Couchbase Lite application:
```
"add" doesn't seem like the right verb here, since you're adding, loading...
"complete" the following for your...?

lamagnifica commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -29,12 +29,22 @@ For experienced developers, this is all you need to add _Couchbase Lite for {par
 That's it! +
 You should be ready to build you app using this version.
-The rest of this content contains more detail, for those who want tp know more about the install or who encountered issues
+The rest of this content contains more detail, for those who want to know more about the install or who encountered issues
```
add period at end?

lamagnifica commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -45,31 +55,59 @@ Couchbase recommends installing Nuget packages via _PackageReference_.
 
+=== Installing Vector Search 
+
+Couchbase recommends installing Nuget packages via _PackageReference_.
```
NuGet (?)

lamagnifica replied:
also, italics are bad according to the style guide

lamagnifica commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -45,31 +55,59 @@ Couchbase recommends installing Nuget packages via _PackageReference_.
+To use Vector Search, you must have Couchbase Lite installed and add the Vector Search extension to your Couchbase Lite application.
+Vector Search is available only for 64-bit architectures and
+Intel processors that support the Advanced Vector Extensions 2 (AVX2) instruction set.
+To verify whether your device supports the AVX2 instructions set, https://www.intel.com/content/www/us/en/support/articles/000090473/processors/intel-core-processors.html[follow these instructions.] 
```
same singular issue

lamagnifica commented on modules/csharp/pages/gs-install.adoc:
```diff
@@ -45,31 +55,59 @@ Couchbase recommends installing Nuget packages via _PackageReference_.
+--
+
+==== Vector Search Package Reference
+This is the recommended method of dependency management for the Vector Search extension because it supports the strict version requirement between the core _Couchbase Lite for .Net_ package and its dependent Support library.
```
this is a little hard to follow. Maybe,

Due to the strict version requirement between the core Couchbase Lite for .Net package and its dependent Support library, Couchbase recommends the following dependency management procedure for the Vector Management Extension.
To manage your dependencies:

dgnicholson commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+You can xref:csharp:gs-install.adoc[use Vector Search] to enhance your mobile and edge applications in a variety of use cases, these include:
+
+* Perform Semantic and Similarity Search on the Edge - Any offline-first mobile or IoT application can benefit from rich semantic text capabilites offered by Vector Search to retrieve contextually relevant data and present it to users.
```
Nit:
```suggestion
* Perform Semantic and Similarity Search on the Edge - Any offline-first mobile or IoT application can benefit from rich semantic text capabilities offered by Vector Search to retrieve contextually relevant data and present it to users.
```

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+== About Vector Search
+
+Vector search is a technique to retrieve semantically similar items based on vector embedding representations of the items in a multi-dimensional space.
```
is this Vector search in general, vs. Vector Search which is the Couchbase extension to Lite? or should this be icapped?

lamagnifica replied:
(the changes in capitalization for Vector S/search are distracting in the next few lines)

osfameron reviewed (approved):
Looks ok to me (I didn't do a very in depth dive, can look closer tomorrow if required)

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+Vector Search is an essential component of Generative AI and Predictive AI applications.
+
+Vector search is a sophisticated data retrieval technique that focuses on matching the contextual meanings of search queries and data entries, rather than simple text matching.
+Vectors are represented by arrays of numbers known as an embedding, which are generated by Large Language Models (LLMs) to depict objects such as text, images, and audio. 
```
I think embeddings plural here?

dgnicholson commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+* Low Latency Application Support - You can run searches locally against a local dataset using a local embedded model. This eliminates the network variability and results in more consistent execution speed.
+Even in the case where the model is not embedded within the local device but is deployed at the edge location, the round trip time (RTT) associated with queries can be significantly reduced compared to searches made over the Internet.
+
+* Cost Per Query Reduction - When you have hundreds of thousands of connected clients querying against a cloud-based LLM, the load on cloud model and operational costs of running the cloud based model can be considerably high.
```
```suggestion
* Cost Per Query Reduction - When you have hundreds of thousands of connected clients querying against a cloud-based LLM, the load and operational costs of running the cloud based model can be considerably high.
```
Nit: Maybe some clarification

lamagnifica replied:
cloud model vs. cloud based model? are they the same or two different things?

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+== Applications of Vector Search
+
+You can xref:csharp:gs-install.adoc[use Vector Search] to enhance your mobile and edge applications in a variety of use cases, these include:
```
two sentences:
"...cases. These include:"

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+* Create Recommendation Engines - Vector Search enables the creation of advanced recommendation systems that analyze semantic similarities between items, user behavior, and preferences. 
+This approach delivers personalized recommendations, improving user engagement and satisfaction.
+
+* Enhance the contextual relevance of your applications with Retrieval Augmented Generation (RAG) - RAG is a technique for enhancing the accuracy and reliability of generative AI models with facts fetched from external sources that are contextual to your application, such as an internal vector database.
```
change in capitalization here is distracting as well

dgnicholson commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+----
+
+First, we initialize the `config` object with a new `VectorIndexConfiguration` object intialized with the following parameters:
```
Nit
```suggestion
First, we initialize the `config` object with a new `VectorIndexConfiguration` object initialized with the following parameters:
```

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+* Enhanced Data Privacy on the Edge - By performing Vector Search within the device, personal data and search queries of a sensitive nature do not have to leave the device.
+
+* Low Latency Application Support - You can run searches locally against a local dataset using a local embedded model. This eliminates the network variability and results in more consistent execution speed.
+Even in the case where the model is not embedded within the local device but is deployed at the edge location, the round trip time (RTT) associated with queries can be significantly reduced compared to searches made over the Internet.
```
"Even in a case..." ? indefinite article here?

dgnicholson reviewed (approved):
A few non-blocker nits, but good to go. Mostly just checked to make sure links and includes worked, as well as very high-level spelling/style check.

ElliotFrancisHunter committed:
DOC-12037: Update modules/csharp/pages/vector-search.adoc

Co-authored-by: Daniel Nicholson <6208288+dgnicholson@users.noreply.github.com>

ElliotFrancisHunter committed:
DOC-12037: Update modules/csharp/pages/working-with-vector-search.adoc

Co-authored-by: Daniel Nicholson <6208288+dgnicholson@users.noreply.github.com>

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+=== About Vector Embeddings
+
+Vector embeddings represent the output of a Machine Learning (ML) model as an array of numbers to capture semantic or contextual relationships between data points.
+This representation encodes how a ML model understands the input or inputs provided to it, based on how the model was initially trained and the internal structure of the model.
```
a ML model? an ML model? I dunno

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
+You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
+Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 
+Training is configured to commence automatically on the first Vector Search query when there are enough vectors to be trained based on the `minimum-training-size` configuration.
```
maybe, 

Training begins automatically for the first Vector Search query that meets the minimum necessary number of vector. 
The minimum is set by the `minimum-training-size` configuration parameter.

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+Vector Search for Couchbase Lite supports the following encoding algorithms:
+
+* None - This will return the highest quality results but at high performance and disk space costs
```
"This returns..."
to match your other bullets

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+It splits the vectors into multiple subspaces and performing scalar quantization on each space independently before compression.
+This can produce higher quality results than Scalar Quantization at the cost of greater complexity.
+
+NOTE: Quantizers are algorithmic processes that map input values from a larger set to output values in a smaller set, common quantization processes can include operations such as rounding and truncation.
```
break into two sentences at the comma ?

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+Vector Indexes are used to store and manage vector representations of content in the form of vector embeddings.
+You can use Vector Indexes to efficiently retrieve vectors, similar to a target vector.
+Before use, a Vector Index needs to be trained to compute the centroids and parameters for encoding the vectors. 
```
maybe link centroids to the section where they're defined below?

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+=== About Centroids
+
+Centroids are vectors that function as the center point of a vector cluster within the data set.
+Each vector is then associated to the vector it is closest to by https://en.wikipedia.org/wiki/K-means_clustering[k-means clustering.]
```
it seems like something happens to allow the vectors to pair up like this (the "is then" makes me wonder, what happened?)
Does identifying the centroid have this effect on the other vectors?

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+Centroids are vectors that function as the center point of a vector cluster within the data set.
+Each vector is then associated to the vector it is closest to by https://en.wikipedia.org/wiki/K-means_clustering[k-means clustering.]
+Each Centroid is contained within a bucket along with its associated vectors, the greater amount of Centroids, the greater the potential accuracy of the model.
```
I'd break into two sentences at the first comma.
Each Centroid is contained within a bucket along with its associated vectors. 
The larger the number of Centroids, the greater the potential accuracy of the model.

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+Each Centroid is contained within a bucket along with its associated vectors, the greater amount of Centroids, the greater the potential accuracy of the model.
+This will also incur a longer indexing time.
+
+Choosing Centroids in vector search involves trade-offs that can impact clustering effectiveness and search efficiency. The initial selection of Centroids, the number chosen, and their sensitivity to high dimensionality and outliers affect the quality of vector clustering.
```
aha! did you CHOOSE a centroid? or configure something for them to get chosen for you?

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+Choosing Centroids in vector search involves trade-offs that can impact clustering effectiveness and search efficiency. The initial selection of Centroids, the number chosen, and their sensitivity to high dimensionality and outliers affect the quality of vector clustering.
+
+The general guideline for the amount of Centroids is approximately the square root of the number of documents.
```
amount ==> number

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+
+=== About Dimensions
+
+Vector dimensions describes the amount of numbers in a given vector embedding, commonly known as its' width.
```
does dimensions have to be plural here? "dimensions describes" doesn't scan well
amount ==> number 
no apostrophe after its.

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+=== About Dimensions
+
+Vector dimensions describes the amount of numbers in a given vector embedding, commonly known as its' width.
+The greater the amount of dimensions, the greater accuracy of results, this also results in greater compute and memory costs and a increase in the latency of the search.
```
break into two sentences at the second comma.

lamagnifica commented on modules/csharp/pages/vector-search.adoc:
```diff
@@ -0,0 +1,137 @@
+The greater the amount of dimensions, the greater accuracy of results, this also results in greater compute and memory costs and a increase in the latency of the search.
+Vector dimensions are dependent on the LLM used to generate the Vector Embeddings. 
+
+NOTE: Couchbase Lite supports dimension sizes in the range of `2 - 2048`.
```
I would write out ...range from 2 to 2048.

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+----
+
+First, we initialize the `config` object with a new `VectorIndexConfiguration` object initialized with the following parameters:
```
we --> you

(use 2nd person not 1st)

lamagnifica replied:
throughout

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+* The expression of the data as a vector.
+
+* The width or `dimensions` of the vector index is set to `3`.
```
no formatting on dimensions (unless they have to type that in somewhere?)

lamagnifica replied:
throughout

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+* The width or `dimensions` of the vector index is set to `3`.
+
+* The amount of `centroids` is set to `2`.
```
same

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+* The width or `dimensions` of the vector index is set to `3`.
+
+* The amount of `centroids` is set to `2`.
+This means that there will be two buckets with a single centroid each that gathers together similar vectors.
```
This results in two buckets, each with a single centroid that acts to gather similar vectors together within that bucket.

? not sure I'm getting centroids...!

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+* The amount of `centroids` is set to `2`.
+This means that there will be two buckets with a single centroid each that gathers together similar vectors.
+
+There are also some default values that we did not set with the `VectorIndexConfiguration()` method, these are:
```
change to two sentences at the comma

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+----
+
+NOTE: The number of vectors, the width or dimensions of the vectors and the training size can incur high CPU and memory costs as the size of each variable increases.
```
comma after ...dimensions of the vectors,...

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+You can use two methods to generate vectors in Couchbase Lite:
+
+. You can call a Machine Learning(ML) model, and embed the generated vectors inside the documents.
```
space before (ML)

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+. Next, create a collection variable passing it the contents of a given collection from our `database` object, in our case, the `colors` collection.
+
+. Finally, create a vector index, `colors_index`, on a collection and pass it our configuration.
```
...pass it ^to^ your...

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+----
+
+NOTE: You can use less storage by using the `prediction()` function as the encoded vectors will only be stored in the index. 
+However, the index time will be longer as vector embedding generation is occurring at run time.
```
is occurring at --> occurs at

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+
+|===
+
+NOTE: The default value for the `limit` parameter is 3.
```
why not put this up in the table with the rest of the info about limit?

lamagnifica commented on modules/csharp/pages/working-with-vector-search.adoc:
```diff
@@ -0,0 +1,227 @@
+----
+
+This function performs vector search against a specific vector index identifier for the specified vector expression.
+If the specified index does not exist, an error will occur on creation of the query.
```
If the specified index does not exist, creating the query results in an error.

(I tried to think of how to reword this somewhere above, just kind of got it now. 
Try to minimize the passive/future even though it's wicked hard)

lamagnifica commented on modules/csharp/partials/release-notes/couchbase-mobile-csharp-release-note.3.2.0.adoc:
```diff
@@ -1,29 +1,165 @@
 [#maint-3-2-0]
-== 3.2.0 Beta 1 -- March 2024
```
you don't want to keep the beta 1 release notes?
Oh, I see, they were completely lacking in useful info!

borrrden replied:
There was no beta 1 release for .net

lamagnifica reviewed (commented):
finished the working with topic

lamagnifica commented:
Wow, complicated stuff! Nice job collecting all the info, @ElliotFrancisHunter
